### PR TITLE
Proxy Support

### PR DESCRIPTION
**Pull request by nixj14**
Added support for configuring HttpClientFactory.  Needed for Proxy support.  Also provide further configuration capabilities now that HttpClientFactory is exposed.

### CODE REVIEW

**Comment by nixj14** [comment]
I signed it!

**Comment by nixj14** [comment]
I signed it!

File: FirebaseAdmin/FirebaseAdmin/AppOptions.cs [hunk 1]
```diff
@@ -58,5 +60,10 @@ internal AppOptions(AppOptions options)
         /// JSON.</para>
         /// </summary>
         public string ServiceAccountId { get; set; }
+
+        /// <summary>
+        /// Gets or sets the HttpClientFactory to use when making Firebase requirests.
+        /// </summary>
+        public HttpClientFactory ClientFactory { get; set; } = new HttpClientFactory();
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin/AppOptions.cs]: Lets call it `HttpClientFactory` without shortening the name.

Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin/AppOptions.cs]: As I understand this is C# 6 syntax. But we do target some .NET versions that are before C# 6. I'm not sure if that's a problem. The safe solution here would be to assign the default value in the constructor. WDYT?

Comment (nixj14) [reply FirebaseAdmin/FirebaseAdmin/AppOptions.cs]: I'm happy to change, though I did notice that FirebaseUserManagerArgs uses "ClientFactory" as well.  In the end, I'm indifferent, but would recommend sticking with one naming convention or the other.

Comment (nixj14) [reply FirebaseAdmin/FirebaseAdmin/AppOptions.cs]: Making the changes now.

Comment (hiranya911) [reply FirebaseAdmin/FirebaseAdmin/AppOptions.cs]: We should be consistent with other Google libraries like Google API client, at least in our public API surface: https://github.com/googleapis/google-api-dotnet-client/blob/master/Src/Support/Google.Apis/Services/BaseClientService.cs#L68

+1 to renaming the internal fields too.

File: FirebaseAdmin/FirebaseAdmin/AppOptions.cs [hunk 2]
```diff
@@ -25,6 +26,8 @@ namespace FirebaseAdmin
     /// </summary>
     public sealed class AppOptions
     {
+        private HttpClientFactory httpClientFactory;
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin/AppOptions.cs]: Can we do this instead?

```
public AppOptions()
{
  this.HttpClientFactory = new HttpClientFactory();
}

public HttpClientFactory HttpClientFactory { get; set; }
```

Then in `FirebaseApp` constructor:

```
if (this.options.HttpClientFactory == null)
{
  // throw
}
```

Without an explicit throw, we could accidentally mask an error in user code.

File: FirebaseAdmin/FirebaseAdmin.Tests/FirebaseAppTest.cs [hunk 3]
```diff
@@ -102,6 +104,28 @@ public void CreateAppOptions()
             Assert.Same(credential, copy.Credential);
             Assert.Equal("test-project", copy.ProjectId);
             Assert.Equal("test@service.account", copy.ServiceAccountId);
+            Assert.Equal(typeof(MockHttpClientFactory), copy.HttpClientFactory.GetType());
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin.Tests/FirebaseAppTest.cs]: We can do a stronger `Assert.Same()` check here.

File: FirebaseAdmin/FirebaseAdmin.Tests/FirebaseAppTest.cs [hunk 4]
```diff
@@ -102,6 +104,28 @@ public void CreateAppOptions()
...
+            };
+            var app = FirebaseApp.Create(options);
+            Assert.Equal("[DEFAULT]", app.Name);
+
+            var copy = app.Options;
+            Assert.NotSame(options, copy);
+            Assert.Same(credential, copy.Credential);
+            Assert.Equal("test-project", copy.ProjectId);
+            Assert.Equal("test@service.account", copy.ServiceAccountId);
+            Assert.Equal(typeof(HttpClientFactory), copy.HttpClientFactory.GetType());
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin.Tests/FirebaseAppTest.cs]: Perhaps a not null check is enough here?

Also add a test case to verify that FirebaseApp doesn't allow null values for HttpClientFactory.

File: FirebaseAdmin/FirebaseAdmin.Tests/Messaging/FirebaseMessagingTest.cs [hunk 5]
```diff
@@ -54,6 +54,17 @@ public void GetMessaging()
...
 
+        [Fact]
+        public void GetMessagingWithClientFactory()
+        {
+            var app = FirebaseApp.Create(new AppOptions() { Credential = MockCredential, HttpClientFactory = new HttpClientFactory() }, "MyApp");
+            FirebaseMessaging messaging = FirebaseMessaging.GetMessaging(app);
+            Assert.NotNull(messaging);
+            Assert.Same(messaging, FirebaseMessaging.GetMessaging(app));
+            app.Delete();
+            Assert.Throws<InvalidOperationException>(() => FirebaseMessaging.GetMessaging(app));
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin.Tests/Messaging/FirebaseMessagingTest.cs]: We should try to do a stronger check here. Otherwise a future code change can silently break this functionality. Easiest option would be to inject a `MockHttpClientFactory` via `AppOptions`. Then invoke `FirebaseMessaging.SendAsync()` and see if that factory get called.

File: FirebaseAdmin/FirebaseAdmin.Tests/FirebaseAppTest.cs [hunk 6]
```diff
@@ -104,7 +104,7 @@ public void CreateAppOptions()
             Assert.Same(credential, copy.Credential);
             Assert.Equal("test-project", copy.ProjectId);
             Assert.Equal("test@service.account", copy.ServiceAccountId);
-            Assert.Equal(typeof(MockHttpClientFactory), copy.HttpClientFactory.GetType());
+            Assert.Same(typeof(MockHttpClientFactory), copy.HttpClientFactory.GetType());
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin.Tests/FirebaseAppTest.cs]: `Assert.Same(options.HttpClientFactory, copy.HttpClientFactory);`

**Comment by hiranya911** [comment]
Sorry for the delay. Approval for this API just came through. I will merge it shortly.
